### ISSUE TITLE
Create functional spec

### ISSUE THREAD

**Issue by jawaad-ahmad** [brata-hsdc/brata#96]
Discussed scenarios with Mike last week. Need to capture these on paper. This is a general document across all components.

I vote to put them here in Markdown rather than a doc on SharePoint inside.

Should I do this on a dedicated branch as well?

I was thinking of keeping it with the other docs in the `transition` folder but it's in the MS repo. Recommendations?

**Comment by jawaad-ahmad** [comment]
This document nails down the user experience and will also contain high-level sequence diagrams.

**Comment by DragonmasterJ** [comment]
> I was thinking of keeping it with the other docs in the transition folder but it's in the MS repo. 

I may have been on the wrong branch, but those documents all look MS specific. It should probably go somewhere more general. I think when we decide where it goes, it'll be easy to determine whether we should give it a branch.

**Comment by ellerychan** [comment]
Seems like a good idea to work on it in a separate branch, but I think it will be important to push it to the **master** branch expediently so everyone gets the latest info.

I've made some URL references in comments and wiki docs to things in branches other than **master**, but I think they might get broken if the branch goes away.  Links to things in the **master** branch should be more robust.

An alternative is to put the doc in the wiki, but then it doesn't stay synced with the code.

I started the _transition_ folder in the MS repo because I didn't know where some of the new design docs should go.  Eventually they will become the defacto docs, so I don't expect them to stay in the _transition_ folder.

If you wanted, you could make a _transition_ folder in this repo, too.

**Comment by DragonmasterJ** [comment]
> If you wanted, you could make a transition folder in this repo, too.

That's more of where I was going with my original thought. 

Although, if station code changes and it changes something in the spec, someone from the station then has to pull the brata repo. Does it then make sense to have a bunch of transition folders, just leave that one in the MS and pull out docs when necessary?

**Comment by ellerychan** [comment]
Teams should manage their docs in whatever way is most convenient for them.

Anyone who is going to modify the document will have to go through the _clone/modify/push/pull request_ process for a doc in a code repo.  I guess that is an advantage to putting a doc in a wiki:  the edit cycle is simpler.

Regardless of where the various design documents are, I think it would be valuable to have a central index that links to them all.  I don't know of an automated way to make sure that index stays up-to-date, so it will probably take a little effort on someone's part to provide that oversight.

**Comment by jawaad-ahmad** [comment]
Here is what I have so far. Please feel free to take a look and provide comments now; no need to wait till the pull request.

[Functional Spec](../blob/functional_spec_issue96/transition/functional_spec.md)

I'm not sure if I'd be worried about keeping this one in-sync with the code; I would imagine it to be relatively static compared to some other documents. Still, it's on its own branch for now. You talked me into it.

It's in a new _transition_ folder in this repo; agreed the _transition_ folders won't stay around forever.

The BRATA area is also our general area, so agreed it makes sense to keep the document here. If someone from station introduces a change that ripples into this document, I think it's understood they're responsible for updating the document even if it's in this repository, though that might break the teams we've set up.

Should someone open a dedicated issue for the central index mentioned above?

**Comment by ellerychan** [comment]
I like your [Functional Spec](https://github.com/brata-hsdc/brata/blob/functional_spec_issue96/transition/functional_spec.md).

> Should someone open a dedicated issue for the central index mentioned above?

Seems like a minor thing.  You could either start it yourself or tack it onto one of @DragonmasterJ 's existing issues.  I think we should be striving to reduce the number of open issues.

**Comment by DragonmasterJ** [comment]
If we want the index to be in the Wiki, maybe tack it on to #88. Then put it back into work so we don't forget it.

**Comment by ellerychan** [comment]
I created the document [here](https://github.com/brata-hsdc/brata/wiki/Design-Document-Index).

**Comment by jawaad-ahmad** [comment]
> I created the document here.

It's empty for now. It still needs to be filled out. I renamed it to _Design Document Index_. _Document Index_ implies all our documents get linked there, which would be useless because the sidebar already does this.

I added a note to Issue #88.

> I think we should be striving to reduce the number of open issues.

I thought we do that by working them. :)

I don't have any problem with open issues. I think we should be striving to keep the number of open `type:1-defect` issues as low as possible, but the number of open issues should be allowed to grow as long as we maintain the list and keep it organized using appropriate labels (e.g. the `priority` labels).

Ideally each issue should tackle only one problem. That makes issues easier to find while searching and prevents items from getting missed because they were just two words in a long-running conversation.

If we're using the issue tracker as an all-encompassing tracker that includes conversations for a discussion board as well, then I would certainly expect the number of issues to continue to grow.

Plus I find it easier to have smaller issues for smaller change sets so that each small issue can be worked off and closed more quickly and more easily (ideally taking up less than 16 hours per issue). Since no one will spend 40 hours/week on this, a large number of smaller issues is even more important here because we can get a more accurate picture of how much work is truly left.

Finally, notes about hyperlinks. General absolute and relative path hyperlinks use the bracket/paren notation:
- [Google](http://www.google.com) `[Google](http://www.google.com)`
- [Design Index](../wiki/Document-Index) `[Design Index](../wiki/Design-Document-Index)`

Links to issues do not:
- Issue #88 in current tracker `Issue #88`
- Station Issue brata-hsdc/brata.station#12 `brata-hsdc/brata.station#12`

> maybe tack it on to `[Organize GitHub Wiki](#88)`

I corrected the issue link for this already.

**Comment by ellerychan** [comment]
Sorry.  It seemed a bit ridiculous that we were starting into a big discussion about whether we should open an issue for performing a simple operation or whether we should add it to an existing issue, when it would be much more expedient to just do it.

**Comment by ellerychan** [comment]
@jawaad-ahmad What do you recommend we do to link from the [Design Document Index](../wiki/Design-Document-Index) to docs under development, like for example: https://github.com/brata-hsdc/brata.masterserver/blob/first_rest_thread_issue13/transition/ms_message_summary.md

This is in a branch I am working on and probably will be for a while, but it would be useful for people to be able to look at the latest version of the document.

Should we push the items from the **transition** folder to the **master** branch more frequently, and link to them there?  If we make the link to my branch version, eventually the link will break.

Can we allow people to push to the **transition** folder without going through a _Pull Request_?

**Comment by jawaad-ahmad** [comment]
I don't think we need the overhead of pushing to `master` or even to `hsdc2016` early just to get people to see. I think we can achieve what we want to do via hyperlinks.

I'm not sure if the link will eventually break, but I do think the current link will eventually become obsolete.

What if we mark the current hyperlink as something that's in-development?

The [Design Document Index](../wiki/Design-Document-Index) looks like it's right on the wiki at this URL in BRATA:

```
https://github.com/brata-hsdc/brata/wiki/Design-Document-Index
```

The [Message Summary](../../brata.masterserver/blob/first_rest_thread_issue13/transition/ms_message_summary.md) is at this URL in a separate repository:

```
https://github.com/brata-hsdc/brata.masterserver/blob/first_rest_thread_issue13/transition/ms_message_summary.md
```

On the [Design Document Index](../wiki/Design-Document-Index), I just added a hyperlink to the [Message Summary](../../brata.masterserver/blob/first_rest_thread_issue13/transition/ms_message_summary.md):

```
* Master Server Message Summary ([in-development](../../../brata.masterserver/blob/first_rest_thread_issue13/transition/ms_message_summary.md))
```

Whenever Issue #13 gets merged back to `hsdc2016`, I think the hyperlink for the Issue #13 version will continue to work, so we can always get to it.

As these docs start getting merged to `hsdc2016`, we can change the entry on the index:

```
* [Master Server Message Summary](../path/to/hsdc2016/version/of/ms_message_summary.md)
```

Thoughts?

**Comment by ellerychan** [comment]
Maybe that's the best we can do.  The problem is that development will continue in another branch, so that link will no longer be current.  I somewhat favor pointing to the one in `hsdc2016` as being the stable, delivered version; and it will get updated when new branches are delivered.

Initially, we will have docs that are evolving rapidly and it might make sense to point to them in another branch so people can view them even before they are pushed to `hsdc2016`, so I like your _in-development_ tag idea.

This is starting to get to my threshold of simple vs. complex process.  It would be one thing if it were everyone's fulltime process, but as it is not, it is another set of rules to remember that will not doubt be applied inconsistently, so we will just need to be tolerant of that.  (I say this knowing that I will be one of the inconsistent ones.)

**Comment by ellerychan** [comment]
> Finally, notes about hyperlinks. General absolute and relative path hyperlinks use the bracket/paren notation:
> 
>    Google `[Google](http://www.google.com)`
>    Design Index `[Design Index](../wiki/Design-Document-Index)`

Be careful when making relative links.  If you check them in the Preview pane of the Wiki composition field, they may not work the same as when they are actually posted, because the URL of the compostion page may be different from the URL of the resulting Wiki page.

You should check the links after the page has been posted, and Edit to fix.
